app/com/linkedin/drelephant/spark/heuristics/ConfigurationUtils.scala

```diff
@@ -0,0 +1,104 @@
+  val SPARK_RESERVED_MEMORY = 300L << 20
+
+  // number of milliseconds in a minute
+  val MILLIS_PER_MIN = 1000D * 60.0D
+
+  // the index for the median value for executor and task metrics distributions
+  val DISTRIBUTION_MEDIAN_IDX = 2
+
+  // the index for the max value for executor and task metrics distributions
+  val DISTRIBUTION_MAX_IDX = 4
+
+  // keys for finding Dr. Elephant configuration parameter values
+  val SPARK_STAGE_EXECUTION_MEMORY_SPILL_THRESHOLD_KEY = "spark_stage_execution_memory_spill_threshold"
+  val SPARK_STAGE_TASK_SKEW_THRESHOLD_KEY = "spark_stage_task_skew_threshold"
+  val SPARK_STAGE_TASK_DURATION_THRESHOLD_KEY = "spark_stage_task_duration_threshold"
+  val SPARK_STAGE_MAX_DATA_PROCESSED_THRESHOLD_KEY = "spark_stage_task_duration_threshold"
+  val TASK_FAILURE_RATE_SEVERITY_THRESHOLDS_KEY = "stage_task_failure_rate_severity_threshold"
+  val MAX_DATA_PROCESSED_THRESHOLD_KEY = "execution_memory_spill_max_data_threshold"
+  val LONG_TASK_TO_STAGE_DURATION_RATIO_KEY = "task_skew_task_to_stage_duration_ratio"
+  val TASK_SKEW_TASK_DURATION_MIN_THRESHOLD_KEY = "task_skew_task_duration_threshold"
+  val MAX_RECOMMENDED_PARTITIONS_KEY = "max_recommended_partitions"
+
+
+  // Severity hresholds for task duration in minutes, when checking to see if the median task
+  // run time is too long for a stage.
+  val DEFAULT_TASK_DURATION_THRESHOLDS =
+    SeverityThresholds(low = 2.5D * MILLIS_PER_MIN, moderate = 5.0D * MILLIS_PER_MIN,
+      severe = 10.0D * MILLIS_PER_MIN, critical = 15.0D * MILLIS_PER_MIN, ascending = true)
+
+  // Severity thresholds for checking task skew, ratio of maximum to median task run times.
+  val DEFAULT_TASK_SKEW_THRESHOLDS =
+    SeverityThresholds(low = 2, moderate = 4, severe = 8, critical = 16, ascending = true)
+
+  // Severity thresholds for checking execution memory spill, ratio of execution spill compared
+  // to the maximum amount of data (input, output, shuffle read, or shuffle write) processed.
+  val DEFAULT_EXECUTION_MEMORY_SPILL_THRESHOLDS =
+    SeverityThresholds(low = 0.01D, moderate = 0.1D, severe = 0.25D, critical = 0.5D, ascending = true)
+
+  // The ascending severity thresholds for the ratio of JVM GC time and task run time,
+  // checking if too much time is being spent in GC.
+  val DEFAULT_GC_SEVERITY_A_THRESHOLDS =
+    SeverityThresholds(low = 0.08D, moderate = 0.09D, severe = 0.1D, critical = 0.15D, ascending = true)
+
+  /** The default severity thresholds for the rate of a stage's tasks failing. */
+  val DEFAULT_TASK_FAILURE_RATE_SEVERITY_THRESHOLDS =
+    SeverityThresholds(low = 0.05D, moderate = 0.1D, severe = 0.15D, critical = 0.2D, ascending = true)
+
+  // The default threshold (3TB) for checking for maximum amount of data processed, for which to
+  // alert for execution memory spill. Tasks processing more data would be expected to have some
+  // amount of spill, due to the large amount of data processed.
+  val DEFAULT_MAX_DATA_PROCESSED_THRESHOLD = "3TB"
```

Comment:
What's the rational behind 3TB as threshold ?

Reply:
If there is a lot of data being processed, it may not be possible (or at least desirable) to avoid execution memory spill. We don't want to recommend increasing executor memory too high (resource consumption), or too many partitions (shuffle overhead), or reduce cores too low (decreased parallelism). For 3TB number, using some reasonable numbers for configuration parameters, I'm estimating based on:

spark.executor.memory / spark.executor.cores * spark.memory.fraction * spark.memory.storageFraction * spark.sql.shuffle.partitions

5GB  / 2 * 0.6 * 0.5 * 4000

Looking at this again though, it would make more sense to calculate the estimate based on some of the other constants and thresholds as well, so that this number would be adjusted automatically with the other values.

Reply:
i think for now just adding more comments about calculating a value should be good and monitories the behaviour.

Reply:
Added some comments.